src/Aspire.Dashboard/Components/Pages/Resources.razor.cs

```diff
@@ -57,90 +67,55 @@ public partial class Resources : ComponentBase, IAsyncDisposable
 
     private readonly CancellationTokenSource _watchTaskCancellationTokenSource = new();
     private readonly ConcurrentDictionary<string, ResourceViewModel> _resourceByName = new(StringComparers.ResourceName);
-    private readonly ConcurrentDictionary<string, bool> _allResourceTypes = [];
-    private readonly ConcurrentDictionary<string, bool> _visibleResourceTypes = new(StringComparers.ResourceName);
     private readonly HashSet<string> _expandedResourceNames = [];
     private string _filter = "";
-    private bool _isTypeFilterVisible;
+    private bool _isFilterPopupVisible;
     private Task? _resourceSubscriptionTask;
     private bool _isLoading = true;
     private string? _elementIdBeforeDetailsViewOpened;
     private FluentDataGrid<ResourceGridViewModel> _dataGrid = null!;
     private GridColumnManager _manager = null!;
     private int _maxHighlightedCount;
 
-    private bool Filter(ResourceViewModel resource) => _visibleResourceTypes.ContainsKey(resource.ResourceType) && (_filter.Length == 0 || resource.MatchesFilter(_filter)) && !resource.IsHiddenState();
+    // Filters in the resource popup
+    private readonly ConcurrentDictionary<string, bool> _resourceTypesToVisibility = new(StringComparers.ResourceName);
+
+    private readonly ConcurrentDictionary<string, bool> _resourceStatesToVisibility = new(StringComparers.ResourceState);
 
-    private async Task OnAllResourceTypesCheckedChangedAsync(bool? areAllTypesVisible)
+    private readonly ConcurrentDictionary<string, bool> _resourceHealthStatusesToVisibility = new(StringComparer.Ordinal);
+
+    internal static string GetStateOrDefaultText(string? state, IStringLocalizer<Dashboard.Resources.Resources> loc)
     {
-        AreAllTypesVisible = areAllTypesVisible;
-        await _dataGrid.SafeRefreshDataAsync();
+        return !string.IsNullOrEmpty(state) ? state : loc[nameof(Dashboard.Resources.Resources.ResourcesResourceHasNoState)];
     }
 
-    private async Task OnResourceTypeVisibilityChangedAsync(string resourceType, bool isVisible)
+    private bool Filter(ResourceViewModel resource)
     {
-        if (isVisible)
-        {
-            _visibleResourceTypes[resourceType] = true;
-        }
-        else
-        {
-            _visibleResourceTypes.TryRemove(resourceType, out _);
-        }
+        return _resourceTypesToVisibility.TryGetValue(resource.ResourceType, out var typeVisible) && typeVisible
+               && _resourceStatesToVisibility.TryGetValue(GetStateOrDefaultText(resource.State, Loc), out var stateVisible) && stateVisible
+                && _resourceHealthStatusesToVisibility.TryGetValue(GetStateOrDefaultText(resource.HealthStatus?.Humanize(), Loc), out var healthStateVisible) && healthStateVisible
```

Comment:
```suggestion
               && _resourceHealthStatusesToVisibility.TryGetValue(GetStateOrDefaultText(resource.HealthStatus?.Humanize(), Loc), out var healthStateVisible) && healthStateVisible
```

Reply:
Have a local method for getting true/false from these dictionaries that avoids the need for TryGetValue.